kai/incident_store.py

```diff
@@ -0,0 +1,292 @@
+__all__ = ["Application", "IncidentStore"]
+
+import copy
+import os
+import pprint
+
+import yaml
+
+from kai.report import Report
+
+
+class Application:
+    def __init__(
+        self,
+        name,
+        report,
+        repo=None,
+        initial_branch=None,
+        solved_branch=None,
+        commitId=None,
+        timestamp=None,
+    ):
+        self.name = name
+        self.report = report
+        self.repo = repo
+        self.initial_branch = initial_branch
+        self.solved_branch = solved_branch
+        self.commitId = commitId
+        self.timestamp = timestamp
+
+
+class IncidentStore:
+    def __init__(self):
+        # applications, keyed on App Name
+        self.applications = {}
+        # cached_violations is defined in comments in self._update_cached_violations
+        self.cached_violations = {}
+        self.pp = pprint.PrettyPrinter(indent=2)
+
+    def add_app_to_incident_store(self, app_name, yaml):
+        r = Report(yaml).get_report()
+        a = Application(name=app_name, report=r)
+        self.applications[app_name] = a
+        self._update_cached_violations(a)
+
+    def get_app_from_incident_store(self, app_name):
+        return self.applications[app_name]
+
+    def get_app_names_from_incident_store(self):
+        return self.applications.keys()
+
+    # todo query hub api for the application
+    def get_app_from_konveyor_hub(self, name):
+        # query hub api for the application
+        # return the application
+        if self.applications[name] is None:
+            return None
+        else:
+            app = self.applications[name]
+            # todo: query hub api for the application
+            # populate the application with the hub api data
+            return app
+
+    # determine if new application analysis report is available
+    def is_new_analysis_report_available(self, application):
+        # query hub api for the application
+        # if the timestamp is newer than the current timestamp
+        # return true
+        # else return false
+        app1 = self.get_app_from_konveyor_hub(application.name)
+        if app1 is not None:
+            if app1.timestamp > application.timestamp:
+                return True
+        return False
+
+    def _update_cached_violations(self, a):
+        """
+        Update the cached_violations with the new application
+        The desired structure
+         cached_violations: {
+           ruleset_name: {
+               violation_name: {
+                   app_name: {
+                       file_path: [ {
+                            variables: {}
```

Comment:
minor 'ni':  May be worth calling out we have 'message' in this as well for future folks who come back to look at comment.

Reply:
I went through a few samples and message field is a part of variables. By calling out, do we want to bring it out of variables and have separate field or just a mechanism to make sure that a 'message' is available for this incident. When I say mechasim, I am thinking of a simple helper function that can quickly lookup the incident and tell me if there is a message or not.

IMO, the message field variables will be very useful in building prompts.

Reply:
I made an assumption that this data structure mapped to what you generate when you run ` ./run_cli.py load kitchensink,eap-coolstore-monolith` and write to samples/incident_datastore

Maybe a bad assumption on my part.  I saw how you had 'message' in the cached data and thought that was helpful.  When I looked at it, it seemed like message was on same level as 'line_number' and 'variables', so I thought it would help to update this comment for future folks looking through code.


```
      src/main/resources/META-INF/resources/bower_components/angular-patternfly/dist/docs/img/RH_atomic.svg:
      - line_number: 4
        message: '`beans_1_1.xsd`: In the root tag, replace the `version` attribute
          value with `3.0`'
        variables:
          matchingText: version="1.1"
      src/main/resources/META-INF/resources/bower_components/angular-patternfly/dist/docs/img/bg-navbar-pf-alt.svg:
      - line_number: 4
        message: '`beans_1_1.xsd`: In the root tag, replace the `version` attribute
          value with `3.0`'
        variables:
          matchingText: version="1.1"
```

Reply:
ACK. Got it. Thanks for the clarification. I will update the comment